### PR TITLE
Restrict Pandas merge suffixes param type to list/tuple to avoid interchange in right and left suffix order

### LINKED ISSUE #33740: BUG: pandas merge suffixes accepting `set` can interchange right and left suffix order
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```python
df_b = pd.DataFrame(dict(p=[1,2,3], q=[2,3,4]))
df_a = pd.DataFrame(dict(p=[2,3,4], q=[1,2,3]))
pd.merge(left=df_b, right=df_a, on=['p'], how='outer', suffixes={'_b', '_a'})
```

#### Current output:
  | p | q_a | q_b
-- | -- | -- | --
0 | 1 | 2.0 | NaN
1 | 2 | 3.0 | 1.0
2 | 3 | 4.0 | 2.0
3 | 4 | NaN | 3.0

#### Expected Output:

  | p | q_b | q_a
-- | -- | -- | --
0 |1 | 2.0 | NaN
1 | 2 | 3.0 | 1.0
2 | 3 | 4.0 | 2.0
3 | 4 | NaN | 3.0

#### Problem description
This [line](https://github.com/pandas-dev/pandas/blob/837daf18d480cce18c25844c591c39da19437252/pandas/core/reshape/merge.py#L670
) is causing the issue:
```
lsuf, rsuf = self.suffixes
```
When you unpack set it returns elements in sorted order:
```
l, r = (2, 1); print(l, r) # l=2, r=1
l, r = [2, 1]; print(l, r) # l=2, r=1
l, r = {2, 1}; print(l, r) # l=1, r=2
```

#### Output of ``pd.show_versions()``

<details>
[INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.2.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.14.165-133.209.amzn2.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.3
numpy            : 1.18.3
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None]
</details>

### PR DESCRIPTION
**Pull request by PuneethaPai**
- [x] closes #33740
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @PuneethaPai for the PR. can you explain the changes? why not just validate in `merge` in pandas\core\reshape\merge.py? Can you also explain the benefit of adding list as an acceptable type? A two tuple can be statically type checked, whereas a list requires runtime length validation.

**Comment by PuneethaPai** [comment]
HI @simonjayhawkins ,
- Added list also to acceptable type, as many tests written in [test_merge.py](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/reshape/merge/test_merge.py), we have used both list and tuple for suffixes.
- Had to introduce type checking as people should not pass `set` which is causing change in order, as `set` returns sorted elements when you unpack.
- Added length validation error tests as no one had done that in tests. But this is not really part of bug #33740.
- I can move the validation in merge method it-self before creating `_MergeOperation` object. Do let me know.

> A two tuple can be statically type checked

did you refer to static checking tools for python like pylint, pyflake, etc?
> whereas a list requires runtime length validation.

I think both list/tuple needs length check. I haven't done length checking explicitly. Relying on python to through error. Example:
```python
In [16]: a, b = (1,2,3)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-16-22772806e702> in <module>
----> 1 a, b = (1,2,3)

ValueError: too many values to unpack (expected 2)

In [17]: a, b = [1, 2, 3]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-c702cf699c19> in <module>
----> 1 a, b = [1, 2, 3]

ValueError: too many values to unpack (expected 2)

In [18]: a, b = [1]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-605f28859c6e> in <module>
----> 1 a, b = [1]

ValueError: not enough values to unpack (expected 2, got 1)
```

I have added tests for these error messages that's all.

**Comment by simonjayhawkins** [comment]
> * Added list also to acceptable type, as many tests written in [test_merge.py](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/reshape/merge/test_merge.py), we have used both list and tuple for suffixes.

the docs clearly state tuple, so I would be ok with changing the tests. need to wait and see what others think.

> * Had to introduce type checking as people should not pass `set` which is causing change in order, as `set` returns sorted elements when you unpack.

again docs clearly state tuple, so could be regarded as a user error, but ok with validating parameters and raising exception to avoid unexpected results.

> * Added length validation error tests as no one had done that in tests. But this is not really part of bug #33740.

sgtm

> * I can move the validation in merge method it-self before creating `_MergeOperation` object. Do let me know.

it seems logical to me minimise the changes to validate prior to the code removed, L670 in pandas/core/reshape/merge.py 

The other changes do reduce the number of parameters passed around, but is that cleanup required here. My preference is to keep PRs atomic where possible. again wait and see what others think.



> > A two tuple can be statically type checked
> 
> did you refer to static checking tools for python like pylint, pyflake, etc?

thinking mypy ( and end users using mypy)

> I think both list/tuple needs length check. I haven't done length checking explicitly. Relying on python to through error. Example:

sgtm.

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -2067,17 +2063,26 @@ def _validate_operand(obj: FrameOrSeries) -> "DataFrame":
         )
 
 
-def _items_overlap_with_suffix(left: Index, lsuffix, right: Index, rsuffix):
+def _items_overlap_with_suffix(left: Index, right: Index, suffixes):
     """
+    Suffixes type validatoin.
```
Comment (jreback) [inline pandas/core/reshape/merge.py:2068]: typo

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -2067,17 +2063,26 @@ def _validate_operand(obj: FrameOrSeries) -> "DataFrame":
...
+
     If two indices overlap, add suffixes to overlapping entries.
 
     If corresponding suffix is empty, the entry is simply converted to string.
 
     """
+    if not isinstance(suffixes, (tuple, list)):
```
Comment (jreback) [inline pandas/core/reshape/merge.py:2075]: let's go back to only allowing tuple

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -2067,17 +2063,26 @@ def _validate_operand(obj: FrameOrSeries) -> "DataFrame":
         )
 
 
-def _items_overlap_with_suffix(left: Index, lsuffix, right: Index, rsuffix):
+def _items_overlap_with_suffix(left: Index, right: Index, suffixes):
```
Comment (jreback) [inline pandas/core/reshape/merge.py:2066]: type suffixes Tuple[str, str]

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -2067,17 +2063,26 @@ def _validate_operand(obj: FrameOrSeries) -> "DataFrame":
...
+    Suffixes type validatoin.
+
     If two indices overlap, add suffixes to overlapping entries.
 
     If corresponding suffix is empty, the entry is simply converted to string.
 
     """
+    if not isinstance(suffixes, (tuple, list)):
+        raise TypeError(
+            f"suffixes should be of type list/tuple. But got {type(suffixes)}"
```
Comment (jreback) [inline pandas/core/reshape/merge.py:2077]: also check that the len is exactly 2

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py (hunk 2 above)
Comment (PuneethaPai) [reply pandas/core/reshape/merge.py:2075]: Sure, So will rewrite test using list to use tuple then,

File: pandas/core/reshape/merge.py (hunk 4 above)
Comment (PuneethaPai) [reply pandas/core/reshape/merge.py]: I have validated for type `tuple`. Should i do explicit validation of `tuple(str, str)`??

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (PuneethaPai) [reply pandas/core/reshape/merge.py:2066]: In my latest pull request, I have validated type for `tuple`. Should I do explicit validation for type to be `tuple(str, str)`??

Comment (simonjayhawkins) [inline pandas/core/reshape/merge.py:2066]:
```suggestion
def _items_overlap_with_suffix(left: Index, right: Index, suffixes: Tuple[str, str]):
```

**Comment by PuneethaPai** [comment]
@simonjayhawkins  or @jreback ,
> Line 2059:74: E231 missing whitespace after ',' 

flake8 as you can see above asks me to add white space after `,`.

But black doesn't like this.
```bash
black --diff pandas/tests/reshape/merge/test_merge.py 
--- pandas/tests/reshape/merge/test_merge.py    2020-05-20 03:05:31 +0000
+++ pandas/tests/reshape/merge/test_merge.py    2020-05-20 03:14:29.050606 +0000
@@ -2054,11 +2054,11 @@
     tm.assert_frame_equal(result, expected)
 
 
 @pytest.mark.parametrize(
     "col1, col2, suffixes",
-    [("a", "a", (None, None)), ("a", "a", ("", None)), (0, 0, (None, "")), ],
+    [("a", "a", (None, None)), ("a", "a", ("", None)), (0, 0, (None, "")),],
 )
 def test_merge_suffix_error(col1, col2, suffixes):
     # issue: 24782
     a = pd.DataFrame({col1: [1, 2, 3]})
     b = pd.DataFrame({col2: [3, 4, 5]})
reformatted pandas/tests/reshape/merge/test_merge.py
All done! ✨ 🍰 ✨
1 file reformatted.
```

As you can see it will delete white space. Can anyone of you suggest fix for this?
This PR is stuck on this simple issue. 😸

File: pandas/tests/reshape/merge/test_merge.py [hunk 5]
```diff
@@ -2056,13 +2056,7 @@ def test_merge_duplicate_suffix(how, expected):
...
 @pytest.mark.parametrize(
     "col1, col2, suffixes",
-    [
-        ("a", "a", [None, None]),
-        ("a", "a", (None, None)),
-        ("a", "a", ("", None)),
-        (0, 0, [None, None]),
-        (0, 0, (None, "")),
-    ],
+    [("a", "a", (None, None)), ("a", "a", ("", None)), (0, 0, (None, "")),],
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/merge/test_merge.py:2059]:
```suggestion
    [("a", "a", (None, None)), ("a", "a", ("", None)), (0, 0, (None, ""))],
```

File: pandas/core/reshape/merge.py [hunk 6]
```diff
@@ -2067,17 +2063,26 @@ def _validate_operand(obj: FrameOrSeries) -> "DataFrame":
...
+    Suffixes type validation.
+
     If two indices overlap, add suffixes to overlapping entries.
 
     If corresponding suffix is empty, the entry is simply converted to string.
 
     """
+    if not isinstance(suffixes, tuple):
+        raise TypeError(
+            f"suffixes should be tuple of (str, str). But got {type(suffixes)}"
```
Comment (simonjayhawkins) [inline pandas/core/reshape/merge.py:2077]:
```suggestion
            f"suffixes should be tuple of (str, str). But got {type(suffixes).__name__}"
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @PuneethaPai for working on this. generally lgtm.

There is a ci failure in the doc build

in doc\source\user_guide\merging.rst

```
result = pd.merge(left, right, on='k', suffixes=['_l', '_r'])
```

needs to be changed also

**Comment by PuneethaPai** [comment]
Thanks @simonjayhawkins . I wasn't able to debug the issue as I couldn't run that ci stage in my local. Thanks for identifying bug.

**Comment by simonjayhawkins** [comment]
@PuneethaPai I _think_ we are going to need a note in the Backwards incompatible API changes
section of doc\source\whatsnew\v1.1.0.rst, something along the line that the parameter is now checked and an exception raised, whereas before a list or set were accepted and that the set could produce unexpected results and add the issue number.

**Comment by PuneethaPai** [comment]
@simonjayhawkins or @jreback sorry to bother again. 
I have updated the v1.1.0 rst file. Not sure how to address this build failure.

Thanks in advance

**Comment by PuneethaPai** [comment]
@simonjayhawkins or @jreback sorry to bother again.
I have refreshed the pull request to resolve merge conflict. Hope it's all good now.
Sorry for multiple to-and-fro conversation. Appreciate your support.

Thanks

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @PuneethaPai nice!

**Comment by jorisvandenbossche** [comment]
This is breaking GeoPandas (where we are using list and not tuple). 

> the docs clearly state tuple

That's not *fully* correct. The `merge` docstring indeed say tuple, but eg `merge_asof` still says "tuple or list". And in our user guide, we were ourselves using a list.

In general, I am certainly fine with restricting the set of accepted types (eg to avoid the confusing issue with sets, the original report), but this is also a change that could easily be done with a deprecation warning. 
And given our versioning policy, if it is easy to do a change with a deprecation, then we should try to do that, IMO.

**Pull request merged**